File: src/projects/bpt/python/BinaryPatriciaTrie.py
'''
Created on Jun 21, 2019

@author: Ares Shackleford
'''
class Node():
    def __init__(self, keyRef, isKey):
        self.left = None
        self.right = None
        self.keyRef = keyRef
        self.isKey = isKey
"""
<p>{@link BinaryPatriciaTrie} is a Patricia Trie over the binary alphabet &#123;      
0, 1 &#125;. By restricting themselves to this small but terrifically useful alphabet, 
Binary Patricia Tries combine all the positive aspects of Patricia Tries while shedding 
the storage cost typically associated with Tries that deal with huge alphabets.</p>
"""   
class BinaryPatriciaTrie():
    
    """
    Simple constructor that will initialize the internals of this. 
    """
    def __init__(self):
        self.__size = 0
        self.__root = Node("", False)
    
    """
    Searches the trie for a given key.
    @param key The input String key.
    @return true if and only if key is in the trie, false otherwise.
    """
    def search(self, key):
        return self.__search_aux(self.__root, key)

    def __search_aux(self, n, key):
        if n is None or len(n.keyRef) > len(key):
            return False
        elif len(n.keyRef) == len(key):
            if n.isKey and n.keyRef == key:
                return True
            return False
        else:
            prefix = self.__compute_pre(key, n.keyRef)
            
            if key[len(prefix)] == "0":
                return self.__search_aux(n.left, key[len(prefix):])
            return self.__search_aux(n.right, key[len(prefix):]) 
    """
    Inserts key into the trie. 
    @param key The input String key. 
    @return true if and only if the key was not already in the trie, false otherwise. 
    """
    def insert(self, key):
        if self.search(key):
            return False
        
        self.__root = self.__insert_aux(self.__root, key)
        self.__size += 1
        
        return True
        
    def __insert_aux(self, n, key):
        if n is None:
            n = Node(key, True)
        elif n.keyRef == key:
            n.isKey = True
        elif n.keyRef == self.__compute_pre(key, n.keyRef):
            if key[len(n.keyRef)] == "0":
                n.left = self.__insert_aux(n.left, key[len(n.keyRef):])
            else:
                n.right = self.__insert_aux(n.right, key[len(n.keyRef):])
        elif key == self.__compute_pre(key, n.keyRef):
            last = n.keyRef[len(key):]
            parent = Node(key,True)
            n.keyRef = last 
            
            if last[0] == "0":
                parent.left = n
            else:
                parent.right = n
            n = parent
        else:
            parent = Node(self.__compute_pre(key, n.keyRef), False)      
            n.keyRef = n.keyRef[len(parent.keyRef):]  
                 
            if n.keyRef[0] == "0":
                parent.left = n
            else:
                parent.right = n
            
            self.__insert_aux(parent, key)
            n = parent
        return n        
        
    """
    Deletes key from the trie. 
    @param key The String key to be deleted.
    @return True if and only if key was contained by the trie before we attempted deletion, 
    false otherwise. 
    """
    def delete(self, key):
        if not self.search(key):
            return False
        
        self.__root = self.__delete_aux(self.__root, key)
        self.__size -= 1
        
        self.__fix_Junk(self.__root)
        return True
    
    def __delete_aux(self, n, key):
        if n.keyRef == key:
            if n.left != None and n.right != None:
                n.isKey = False
            elif n.left is None and n.right is None:
                n = None
            else:
                if n.left != None:
                    self.__merge(n,n.left)
                else:
                    self.__merge(n,n.right)   
                    
        elif n.keyRef == self.__compute_pre(key, n.keyRef):
            if key[len(n.keyRef)] == "0":
                n.left = self.__delete_aux(n.left, key[len(n.keyRef):])
            else:
                n.right = self.__delete_aux(n.right, key[len(n.keyRef):])
        return n
    
    def __compute_pre(self, key, ref):
        pre = ""
        
        if len(key) > len(ref):
            length = len(ref)
        else:
            length = len(key)
        
        for i in range(length):
            if key[i] == ref[i]:
                pre += key[i]
            else:
                break
    
        return pre
    
    def __merge(self, parent, child):
        if child is not None:
            parent.keyRef += child.keyRef
            parent.isKey = child.isKey
            parent.left = child.left
            parent.right = child.right
        else:
            raise AssertionError("__merge(): cannot have a null child")
        
    def __fix_Junk(self, n):
        if n is not None:
            self.__fix_Junk(n.left)
            if n.isKey is False and (n.left is None or n.right is None) and n != self.__root:
                if n.right is None:
                    self.__merge(n, n.left)
                else:
                    self.__merge(n, n.right)
            self.__fix_Junk(n.right)    
    """
    Queries the trie for emptiness. 
    @return true if and only if {@link #getSize()} == 0, false otherwise. 
    """
    def is_Empty(self):
        return self.__size == 0
    
    """
    Returns the number of keys in the tree. 
    @return The number of keys in the tree. 
    """
    def get_Size(self):
        return self.__size
    
    """
    <p>Performs an <i>inorder (symmetric) traversal</i> of the Binary Patricia Trie. 
    Remember from lecture that inorder traversal in tries is NOT sorted traversal, 
    unless all the stored keys have the same length. This is of course not required by your 
    implementation, so you should make sure that in your tests you are not expecting this 
    method to return keys in lexicographic order. We put this method in the interface because it 
    helps us test your submission thoroughly and it helps you debug your code! </p> 
    
    <p>We <b>neither require nor test </b> whether the {@link Iterator} returned by this 
    method is fail-safe or fail-fast. This means that you  do <b>not</b> need to test for 
    thrown {@link java.util.ConcurrentModificationException}s and we do <b>not</b> test your 
    code for the possible occurrence of concurrent modifications.</p> 
    
    <p>We also assume that the {@link Iterator} is <em>immutable</em>, i,e we do <b>not</b> 
    test for the behavior of {@link Iterator#remove()}. You can handle it any way you want 
    for your own application, yet <b>we</b> will <b>not</b> test for it.</p> 
    
    @return An {@link Iterator} over the {@link String} keys stored in the trie, 
    exposing the elements in <i>symmetric order</i>. 
    """
    def in_Order_Traversal(self):
        words = []
        
        self.__in_Order(self.__root, words, "")
        
        return words.__iter__()
    
    def __in_Order(self, n, words, word):
        if n != None:
            word += n.keyRef
            self.__in_Order(n.left, words, word)
            if n.isKey:
                words.append(word)
            self.__in_Order(n.right, words, word)
    """
    Finds the longest {@link String} stored in the Binary Patricia Trie. 
    
    @return <p>The longest {@link String} stored in this. If the trie is empty, 
    the empty string &quot;&quot; should be returned. Careful: the empty string 
    &quot;&quot;is <b>not</b> the same string as &quot; &quot;; the latter is a 
    string consisting of a single <b>space character</b>! It is also <b>not the 
    same as the</b> null <b>reference</b>!</p> 
    
    <p>Ties should be broken in terms of <b>value</b> of the bit string. For example, 
    if our trie contained only the binary strings 01 and 11, <b>11</b> would be the 
    longest string. If our trie contained only 001 and 010, <b>010</b> would be the 
    longest string.</p>
    """
    def get_Longest(self):
        length = 0
        word = ""
        
        if self.is_Empty():
            return self.__root.keyRef
        
        for i in self.in_Order_Traversal():
            if len(i) > length:
                word = i
                length = len(word)
            elif len(i) == length:
                if int(i, 2) > int(word,2):
                    word = i
        return word
    """
    Makes sure that your trie doesn't have splitter nodes with a single child. 
    In a Patricia trie, those nodes should be pruned. Be careful with the 
    implementation of this method, since our tests call it to make sure your 
    deletions work correctly! That is to say, if your deletions work well, 
    but you have made an error in this (far easier) method, you will <b>still</b> 
    not be passing our tests!
    
    @return true iff all nodes in the trie either denote stored strings or split 
    into two subtrees, false otherwise.
    """
    def is_Junk_Free(self):
        return self.__junk(self.__root.left) and self.__junk(self.__root.right)
    
    def __junk(self, n):
        if n is None:
            return True
        else:
            if n.isKey == False and (n.left is None or n.right is None):
                return False
            else:
                return self.__junk(n.left) and self.__junk(n.right)            

File: src/projects/bpt/java/BinaryPatriciaTrie.java
package projects.bpt.java;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * <p>BinaryPatriciaTrie is a Patricia Trie over the binary alphabet &#123;	 0, 1 &#125;. By restricting themselves
 * to this small but terrifically useful alphabet, Binary Patricia Tries combine all the positive
 * aspects of Patricia Tries while shedding the storage cost typically associated with Tries that
 * deal with huge alphabets.</p>
 *
 * @author <a href = "https://github.com/JasonFil">Jason Filippou</a> &amp; <a href ="mailto:fyang623@gmail.com">Fan Yang</a>
 */
public class BinaryPatriciaTrie {

    private class TrieNode {

        private TrieNode left, right;
        private String str;

        // isKey indicates whether the node is a key node or not, i.e. whether the string achieved by
        // concatenating the path from the root to this node is a key or not

        private boolean isKey;

        TrieNode() {
            str = "";
            isKey = false;
        }

        TrieNode(String str, boolean isKey) {
            this.str = str;
            this.isKey = isKey;
        }
    }


    // JASON: Made access private (was package-private).
    private TrieNode root;

    private int size;
    private boolean flag; // a boolean variable that will be used by insert() and delete()

    /**
     * Simple constructor that will initialize the internals of this.
     */

    public BinaryPatriciaTrie() {
        root = new TrieNode();
        flag = false;
        size = 0;
    }

    /**
     * Searches the trie for a given key.
     *
     * @param key The input String key.
     * @return true if and only if key is in the trie, false otherwise.
     */
    public boolean search(String key) {
        return search(root, key);
    }

    /**
     * <p>A helper method used by  {@link #search(String)}. Searches the subtree
     * rooted at curr for key. </p>
     *
     * @param curr The current node.
     * @param key  The key to be searched in the subtree rooted at the current node.
     * @return true if and only if key is in the trie, false otherwise.
     */

    private boolean search(TrieNode curr, String key) {
        if (curr == null || key == null) return false;

        // If curr.str is shorter than the key being searched, we check whether curr.str
        // is a prefix of the key
        if (curr.str.length() < key.length()) {
            // If curr.str is not a prefix of the key, then the key is not in the Patricia trie.
            if (!key.substring(0, curr.str.length()).equals(curr.str))
                return false;
            // If str is a prefix of the key, we recurse into the left or right child.
            if (key.charAt(curr.str.length()) == '0')
                return search(curr.left, key.substring(curr.str.length()));
            else
                return search(curr.right, key.substring(curr.str.length()));
        }

        return curr.str.length() == key.length() && curr.str.equals(key) && curr.isKey;
    }

    /**
     * Inserts key into the trie.
     *
     * @param key The input String key.
     * @return true if and only if the key was not already in the trie, false otherwise.
     */
    public boolean insert(String key) {
        if (key == null || key.trim().equals("")) return false;

        // check whether key is a valid binary string
        for (int i = 0; i < key.trim().length(); i++)
            if (key.charAt(i) != '0' && key.charAt(i) != '1') return false;

        flag = true;
        root = insert(root, key.trim());
        size = flag ? (size + 1) : size;
        return flag;
    }


    /**
     * A helper method used by {@link #insert(String key)}.
     * Inserts key into the subtree of the trie rooted at curr.
     *
     * @param curr The current node.
     * @param key  The key to be added to the subtree rooted at the current node.
     * @return The root of the subtree after the insertion of the key.
     */

    private TrieNode insert(TrieNode curr, String key) {
        // If the current node is a leaf, insert the key
        if (curr == null) {
            curr = new TrieNode(key, true);
        } else if (curr.str.equals(key)) {
            // If the key being searched is equal to the string stored in the current node, we simply
            // set curr.isKey to true and set the value of flag appropriately.
            flag = !curr.isKey; // So flag becomes false if we insert a duplicate.
            curr.isKey = true;
        }

        // If the key being inserted is a prefix of the string stored in the current node, we split
        // the current node to two and set isKey to true in the resulting upper node.
        else if (key.length() < curr.str.length() && curr.str.substring(0, key.length()).equals(key)) {

            TrieNode temp = curr;
            temp.str = temp.str.substring(key.length()); // Begins at index key.length() through the end. Fan builds the string sequentially!
            curr = new TrieNode(key, true);
            if (temp.str.charAt(0) == '0')
                curr.left = temp;
            else
                curr.right = temp;
        }

        // If the string stored in the current node is a prefix of the key being inserted, we recurse
        // into the left or right child, as appropriate.
        else if (key.length() > curr.str.length() && key.substring(0, curr.str.length()).equals(curr.str)) {
            if (key.charAt(curr.str.length()) == '0')
                curr.left = insert(curr.left, key.substring(curr.str.length()));
            else
                curr.right = insert(curr.right, key.substring(curr.str.length()));
        }

        // In all other cases, we split the current node to two and add a second child to the resulting
        // upper node appropriately, as specified in the method splitInsert().
        else
            curr = splitInsert(curr, key);

        return curr;
    }

    /**
     * <p>A helper method used by {@link #insert(TrieNode, String)}. Splits the node
     *  referred to by curr into two new nodes and creates a father node that
     *  contains the <i>longest shared prefix of the {@link String} in curr and key.
     *  Decides which one of the two nodes needs to be the right (resp. left) child of the new father node,
     *  and attaches them to said node. </p>
     *
     * @param curr The current node.
     * @param key  The key to be added to the subtree rooted at the current node.
     * @return The root of the subtree after the split.
     */

    private TrieNode splitInsert(TrieNode curr, String key) {
        int i = getSharedPrefixLength(curr.str, key);
        assert (i < Math.min(curr.str.length(), key.length()));

        TrieNode temp = curr;
        temp.str = temp.str.substring(i);
        curr = new TrieNode(key.substring(0, i), false);
        if (key.charAt(i) == '0') {
            curr.left = new TrieNode(key.substring(i), true);
            curr.right = temp;
        } else {
            curr.right = new TrieNode(key.substring(i), true);
            curr.left = temp;
        }
        return curr;
    }


    /**
     * A helper method used by {@link #splitInsert(TrieNode curr, String key)}.
     *
     * @param s1 Input string &#35;1.
     * @param s2 Input string &#35;2.
     * @return The length of the longest prefix shared by the two input strings.
     */
    private int getSharedPrefixLength(String s1, String s2) {
        int i = s1.length() < s2.length() ? s1.length() : s2.length();

        // Fan goes from the end to the beginning, implementing the logic:
        // As long as the substrings are not exactly equal, decrement the length
        // of the possible shared prefix.

        while (!s1.substring(0, i).equals(s2.substring(0, i)))
            i--;
        return i;
    }

    /**
     * Deletes key from the trie.
     *
     * @param key The String key to be deleted.
     * @return True if and only if key was contained by the trie before we attempted deletion, false otherwise.
     */
    public boolean delete(String key) {
        if (key == null) return false;
        flag = false;
        root = delete(root, key.trim());
        size = flag ? (size - 1) : size;
        if(size == 0) root = new TrieNode(); // For consistency with isEmpty().
        return flag;
    }

    /**
     * <p>A helper method used by {@link #delete(String key)}.
     * Deletes the string key from the subtree rooted at the current node. </p>
     *
     * @param curr The current node.
     * @param key  The {@link String} key to be deleted from the subtree rooted at the current node.
     * @return The root of the subtree after the deletion.
     */

    private TrieNode delete(TrieNode curr, String key) {
        // Current node is null indicates that the key is not found in the trie
        if (curr == null) return null;

        // Base case : the key is found in the current node (4 sub cases)
        if (curr.str.equals(key) && curr.isKey) {
            // sub case #1 : both children are not null. We simply set isKey of the current node to false.
            if (curr.left != null && curr.right != null)
                curr.isKey = false;
            // sub case #2 : only the left child is not null. Merge the current node with its left child.
            else if (curr.left != null) {
                curr.left.str = curr.str + curr.left.str;
                curr = curr.left; // TODO: This is assigning to a local copy that is then returned... fix with a retVal variable.
            }
            // sub case #3 : only the right child is not null. Merge the current node with its right child.
            else if (curr.right != null) {
                curr.right.str = curr.str + curr.right.str;
                curr = curr.right; /// TODO: Same
            }
            // sub case #4 : both children are null. Simply set the current node to null.
            else {
                curr = null;        // TODO: Same
            }
            // The key is successfully deleted, so we set flag to true.
            flag = true;
        }

        // If the string stored in the current node is a prefix of the key to be deleted, we recurse
        // into the left or right child as appropriate.
        else if (curr.str.length() < key.length() && key.substring(0, curr.str.length()).equals(curr.str)) {

            if (key.charAt(curr.str.length()) == '0') {
                curr.left = delete(curr.left, key.substring(curr.str.length()));
                if(!curr.isKey && curr.left == null){
                    if(curr.right != null) {
                        curr.right.str = curr.str + curr.right.str;
                    }
                    curr = curr.right;   // Throw away useless node. // TODO: Same.
                }
            }
            else {
                curr.right = delete(curr.right, key.substring(curr.str.length()));
                if(!curr.isKey && curr.right == null){
                    if(curr.left != null) {
                        curr.left.str = curr.str + curr.left.str;
                    }
                    curr = curr.left;   // Throw away useless node. // TODO: Same
                }
            }
        }

        return curr;
    }

    /**
     * Queries the trie for emptiness.
     *
     * @return true if and only if {@link #getSize()} == 0, false otherwise.
     */
    public boolean isEmpty() {
        return root.left == null && root.right == null;
    }

    /**
     * Returns the number of keys in the tree.
     *
     * @return The number of keys in the tree.
     */
    public int getSize() {
        return size;
    }

    /**
     * <p>Performs an <i>inorder (symmetric) traversal</i> of the Binary Patricia Trie. Remember from lecture that inorder
     * traversal in tries is NOT sorted traversal, unless all the stored keys have the same length. This
     * is of course not required by your implementation, so you should make sure that in your tests you
     * are not expecting this method to return keys in lexicographic order. We put this method in the
     * interface because it helps us test your submission thoroughly and it helps you debug your code! </p>
     *
     * <p>We <b>neither require nor test </b> whether the {@link Iterator} returned by this method is fail-safe or fail-fast.
     * This means that you  do <b>not</b> need to test for thrown {@link java.util.ConcurrentModificationException}s and we do
     * <b>not</b> test your code for the possible occurrence of concurrent modifications.</p>
     *
     * <p>We also assume that the {@link Iterator} is <em>immutable</em>, i,e we do <b>not</b> test for the behavior
     * of {@link Iterator#remove()}. You can handle it any way you want for your own application, yet <b>we</b> will
     * <b>not</b> test for it.</p>
     *
     * @return An {@link Iterator} over the {@link String} keys stored in the trie, exposing the elements in <i>symmetric
     * order</i>.
     */
    public Iterator<String> inorderTraversal() {
        ArrayList<String> list = new ArrayList<>();
        inOrderTraversal(root.left, list, "");
        inOrderTraversal(root.right, list, "");
        return list.iterator();
    }

    private void inOrderTraversal(TrieNode curr, List<String> list, String prefix) {
        if (curr == null) return;
        else inOrderTraversal(curr.left, list, prefix + curr.str);
        if (curr.isKey)
            list.add(prefix + curr.str);
        inOrderTraversal(curr.right, list, prefix + curr.str);
    }

    /**
     * Finds the longest {@link String} stored in the Binary Patricia Trie.
     * @return <p>The longest {@link String} stored in this. If the trie is empty, the empty string &quot;&quot; should be
     * returned. Careful: the empty string &quot;&quot;is <b>not</b> the same string as &quot; &quot;; the latter is a string
     * consisting of a single <b>space character</b>! It is also <b>not the same as the</b> null <b>reference</b>!</p>
     *
     * <p>Ties should be broken in terms of <b>value</b> of the bit string. For example, if our trie contained
     * only the binary strings 01 and 11, <b>11</b> would be the longest string. If our trie contained
     * only 001 and 010, <b>010</b> would be the longest string.</p>
     */
    public String getLongest() {
        String longest = "", temp;
        Iterator<String> iterator = inorderTraversal();
        while (iterator.hasNext()) {
            temp = iterator.next();
            if (temp.length() > longest.length())
                longest = temp;
            else if (temp.length() == longest.length()) {
                int j = 0;
                while (temp.charAt(j) == longest.charAt(j))
                    j++;
                if (temp.charAt(j) == '1')
                    longest = temp;
            }
        }
        return longest;
    }

    /**
     * Makes sure that your trie doesn't have splitter nodes with a single child. In a Patricia trie, those nodes should
     * be pruned. Be careful with the implementation of this method, since our tests call it to make sure your deletions work
     * correctly! That is to say, if your deletions work well, but you have made an error in this (far easier) method,
     * you will <b>still</b> not be passing our tests!
     *
     * @return true iff all nodes in the trie either denote stored strings or split into two subtrees, false otherwise.
     */
    public boolean isJunkFree(){
        return isEmpty() || (isJunkFree(root.left) && isJunkFree(root.right));
    }

    private boolean isJunkFree(TrieNode n){
        if(n == null){   // Null subtrees trivially junk-free
            return true;
        }
        if(!n.isKey){   // Non-key nodes need to be strict splitter nodes
            return ( (n.left != null) && (n.right != null) && isJunkFree(n.left) && isJunkFree(n.right) );
        } else {
            return ( isJunkFree(n.left) && isJunkFree(n.right) ); // But key-containing nodes need not.
        }
    }
}
